File: lwaf_3D/IScaled.java
package lwaf_3D;

import lwaf_core.vec3;

public interface IScaled<R> {
    vec3 getScale();

    R setScale(vec3 scale);

    default R setScale(float x, float y, float z) {
        return setScale(new vec3(x, y, z));
    }

    default R setScale(float scale) {
        return setScale(scale, scale, scale);
    }

    default R scaleBy(vec3 scale) {
        return setScale(getScale().mul(scale));
    }

    default R scaleBy(float x, float y, float z) {
        return scaleBy(new vec3(x, y, z));
    }

    default R scaleBy(float scale) {
        return scaleBy(scale, scale, scale);
    }
}

File: lwaf_3D/IScaled.kt
package lwaf_3D

import lwaf_core.times
import lwaf_core.vec3

interface IScaled<R> {
    val scale: vec3

    fun setScale(scale: vec3): R

    fun setScale(x: Float, y: Float, z: Float): R {
        return setScale(vec3(x, y, z))
    }

    fun setScale(scale: Float): R {
        return setScale(scale, scale, scale)
    }

    fun scaleBy(scale: vec3): R {
        return setScale(this.scale * scale)
    }

    fun scaleBy(x: Float, y: Float, z: Float): R {
        return scaleBy(vec3(x, y, z))
    }

    fun scaleBy(scale: Float): R {
        return scaleBy(scale, scale, scale)
    }
}
